File: ruby/bob/bob.rb
class Bob
  class << self
    def hey(remark)
      return "Whoa, chill out!"   if yelling?(remark)
      return "Sure."              if asking?(remark)
      return "Fine. Be that way!" if silence?(remark)

      "Whatever."
    end

    def yelling?(remark)
      letters = remark.scan(/[^1-9|\W]/)
      return false if letters.empty?

      letters.all? { |letter| letter == letter.upcase }
    end

    def asking?(remark)
      remark.strip.end_with?("?")
    end

    def silence?(remark)
      remark.strip.empty?
    end
  end
end

module BookKeeping
  VERSION = 1
end

File: go/bob/bob.go
package bob

import (
	"fmt"
	"regexp"
)

// Hey returns Bob's answer.
func Hey(remark string) string {
	fmt.Println("remark:", remark)
	fmt.Println("------")

	yelling := regexp.MustCompile(`[A-Z]!|[A-Z]+$`)
	questioning := regexp.MustCompile(`[^A-Z]\?$`)
	forcefulQuestioning := regexp.MustCompile(`[A-Z]\?$`)

	if yelling.MatchString(remark) {
		return "Whoa, chill out!"
	}

	if questioning.MatchString(remark) {
		return "Sure."
	}

	if forcefulQuestioning.MatchString(remark) {
		return "Calm down, I know what I'm doing!"
	}

	return "Whatever."
}
